### PR TITLE
planner: fix constant propagation for PredicatePushDown

### LINKED ISSUE #14481: constant propagation does not work
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
``` sql
CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  `c` int(11) DEFAULT NULL
);
desc select * from t where a = 1 and a = 2;
```

2. What did you expect to see?
``` sql
tidb> desc select * from t where a = 1 and a = 2;

TableDual
```


3. What did you see instead?
``` sql
tidb> desc select * from t where a = 1 and a = 2;
+---------------------+----------+-----------+------------------------------------------------------------+
| id                  | count    | task      | operator info                                              |
+---------------------+----------+-----------+------------------------------------------------------------+
| TableReader_7       | 0.00     | root      | data:Selection_6                                           |
| └─Selection_6       | 0.00     | cop[tikv] | eq(amldb.t.a, 1), eq(amldb.t.a, 2)                         |
|   └─TableScan_5     | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+-----------+------------------------------------------------------------+
3 rows in set (0.00 sec)
```


4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
``` sql
tidb> select tidb_version();
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-alpha-1334-g07e642c92
Git Commit Hash: 07e642c9230ccb7c1537b27442f1fe8433e65f8a
Git Branch: master
UTC Build Time: 2020-01-15 06:52:57
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by gauss1314**
Signed-off-by: gauss <gauss1314@gmail.com>

### What problem does this PR solve? 

This fix #14481  

### What is changed and how it works?

fix constant propagation bug for PredicatePushDown.

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: cmd/explaintest/r/explain_easy.result [hunk 1]
```diff
@@ -450,7 +450,7 @@ drop table if exists t;
 create table t(a bigint primary key);
 explain select * from t where a = 1 and a = 2;
 id	count	task	operator info
-TableDual_5	0.00	root	rows:0
+TableDual_5	8000.00	root	rows:0
```
Comment (winoros) [inline cmd/explaintest/r/explain_easy.result]: It's stange that the row count changed

File: cmd/explaintest/r/explain_easy.result [hunk 2]
```diff
@@ -745,7 +745,7 @@ create table t(a int, b int);
 explain select a, b from (select a, b, avg(b) over (partition by a)as avg_b from t) as tt where a > 10 and b < 10 and a > avg_b;
 id	count	task	operator info
 Projection_8	2666.67	root	test.t.a, test.t.b
-└─Selection_9	2666.67	root	gt(cast(test.t.a), Column#5), lt(test.t.b, 10)
+└─Selection_9	2666.67	root	gt(cast(test.t.a), Column#5), gt(test.t.a, 10), lt(test.t.b, 10)
```
Comment (winoros) [inline cmd/explaintest/r/explain_easy.result]: This means that `a > 10` is wrongly pruned before?

File: planner/core/rule_predicate_push_down.go [hunk 3]
```diff
@@ -72,11 +72,11 @@ func splitSetGetVarFunc(filters []expression.Expression) ([]expression.Expressio
 
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
 func (p *LogicalSelection) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan) {
+	p.Conditions = expression.PropagateConstant(p.ctx, p.Conditions)
```
Comment (winoros) [inline planner/core/rule_predicate_push_down.go]: Why this change can take effect?
I just see that there's `PropagateConstant` in Sel's `PredicatePushDown` and in `addSelection`.

File: cmd/explaintest/r/explain_easy.result (hunk 1 above)
Comment (gauss1314) [reply cmd/explaintest/r/explain_easy.result]: I  don't know why row count was 8000.00, but it just happened when I run `./run-tests -r all ../../bin/tidb-server` .

File: cmd/explaintest/r/explain_easy.result (hunk 2 above)
Comment (gauss1314) [reply cmd/explaintest/r/explain_easy.result]: Same as the before reply.

File: planner/core/rule_predicate_push_down.go (hunk 3 above)
Comment (gauss1314) [reply planner/core/rule_predicate_push_down.go]: I had do some change, what I want to do is do `PropagateConstant`  before `PredicatePushDown` .

File: planner/core/rule_predicate_push_down.go [hunk 4]
```diff
@@ -72,16 +72,18 @@ func splitSetGetVarFunc(filters []expression.Expression) ([]expression.Expressio
 
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
 func (p *LogicalSelection) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan) {
-	canBePushDown, canNotBePushDown := splitSetGetVarFunc(p.Conditions)
-	retConditions, child := p.children[0].PredicatePushDown(append(canBePushDown, predicates...))
+	newConditions := append(predicates, p.Conditions...)
+	newConditions = expression.PropagateConstant(p.ctx, newConditions)
+	// Return table dual when filter is constant false or null.
+	dual := Conds2TableDual(p, p.Conditions)
```
Comment (francis0407) [inline planner/core/rule_predicate_push_down.go]: I think we should use `Conds2TableDual(p, newConditions)` here.

File: cmd/explaintest/r/explain_easy.result (hunk 1 above)
Comment (gauss1314) [reply cmd/explaintest/r/explain_easy.result]: done

File: planner/core/testdata/plan_suite_unexported_in.json [hunk 5]
```diff
@@ -417,48 +417,48 @@
     "name": "TestTablePartition",
     "cases": [
       {
-        "SQL":   "select * from t",
```
Comment (francis0407) [inline planner/core/testdata/plan_suite_unexported_in.json]: This file is not supposed to be changed.

File: planner/core/testdata/integration_suite_in.json [hunk 6]
```diff
@@ -38,9 +38,9 @@
       // Row count of IndexScan should be 2.
       "explain select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a and t1.c = t2.c",
       // Row count of IndexScan should be 2.
-	  "explain select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a and t1.c <= t2.b",
+      "explain select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a and t1.c <= t2.b",
       // Row count of IndexScan should be 1.
-	  "explain select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a and t2.b = 1"
+      "explain select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a and t2.b = 1"
```
Comment (francis0407) [inline planner/core/testdata/integration_suite_in.json]: ditto

File: planner/core/memtable_predicate_extractor_test.go [hunk 7]
```diff
@@ -73,9 +75,15 @@ func (s *extractorSuite) getLogicalMemTable(c *C, se session.Session, parser *pa
...
-	logicalMemTable := leafPlan.(*plannercore.LogicalMemTable)
-	return logicalMemTable
+	_, isDual := leafPlan.(*plannercore.LogicalTableDual)
+	if isDual {
+		return plannercore.LogicalMemTable{
+			DBName:    model.NewCIStr("information_schema"),
+			Extractor: nil,
+			TableInfo: nil,
+		}.Init(mock.NewContext(), 0)
+	}
```
Comment (francis0407) [inline planner/core/memtable_predicate_extractor_test.go]: Can we just return `nil` here?

File: planner/core/memtable_predicate_extractor_test.go [hunk 8]
```diff
@@ -785,15 +797,17 @@ func (s *extractorSuite) TestInspectionResultTableExtractor(c *C) {
...
 		logicalMemTable := s.getLogicalMemTable(c, se, parser, ca.sql)
-		c.Assert(logicalMemTable.Extractor, NotNil)
-
-		clusterConfigExtractor := logicalMemTable.Extractor.(*plannercore.InspectionResultTableExtractor)
-		if len(ca.rules) > 0 {
-			c.Assert(clusterConfigExtractor.Rules, DeepEquals, ca.rules, Commentf("SQL: %v", ca.sql))
-		}
-		if len(ca.items) > 0 {
-			c.Assert(clusterConfigExtractor.Items, DeepEquals, ca.items, Commentf("SQL: %v", ca.sql))
+		if logicalMemTable.Extractor == nil {
```
Comment (francis0407) [inline planner/core/memtable_predicate_extractor_test.go]: And we only need to check whether `logicalMemTable == nil` ?

File: planner/core/memtable_predicate_extractor_test.go [hunk 9]
```diff
@@ -235,12 +237,14 @@ func (s *extractorSuite) TestClusterConfigTableExtractor(c *C) {
 	}
 	for _, ca := range cases {
 		logicalMemTable := s.getLogicalMemTable(c, se, parser, ca.sql)
-		c.Assert(logicalMemTable.Extractor, NotNil)
-
-		clusterConfigExtractor := logicalMemTable.Extractor.(*plannercore.ClusterTableExtractor)
-		c.Assert(clusterConfigExtractor.NodeTypes, DeepEquals, ca.nodeTypes, Commentf("SQL: %v", ca.sql))
-		c.Assert(clusterConfigExtractor.Instances, DeepEquals, ca.instances, Commentf("SQL: %v", ca.sql))
-		c.Assert(clusterConfigExtractor.SkipRequest, DeepEquals, ca.skipRequest, Commentf("SQL: %v", ca.sql))
+		if logicalMemTable.Extractor == nil {
```
Comment (francis0407) [inline planner/core/memtable_predicate_extractor_test.go]: Need to be changed here?

File: planner/core/memtable_predicate_extractor_test.go [hunk 10]
```diff
@@ -515,21 +519,23 @@ func (s *extractorSuite) TestClusterLogTableExtractor(c *C) {
...
-		if ca.startTime > 0 {
-			c.Assert(clusterConfigExtractor.StartTime, Equals, ca.startTime, Commentf("SQL: %v", ca.sql))
-		}
-		if ca.endTime > 0 {
-			c.Assert(clusterConfigExtractor.EndTime, Equals, ca.endTime, Commentf("SQL: %v", ca.sql))
-		}
-		c.Assert(clusterConfigExtractor.Patterns, DeepEquals, ca.patterns, Commentf("SQL: %v", ca.sql))
-		if len(ca.level) > 0 {
-			c.Assert(clusterConfigExtractor.LogLevels, DeepEquals, ca.level, Commentf("SQL: %v", ca.sql))
+		if logicalMemTable.Extractor == nil {
```
Comment (francis0407) [inline planner/core/memtable_predicate_extractor_test.go]: ditto

**Comment by francis0407** [comment]
It seems this PR affects the `prepared plan`, I'm not quite sure about it. PTAL @eurekaka @winoros

File: planner/core/rule_predicate_push_down.go [hunk 11]
```diff
@@ -73,15 +73,17 @@ func splitSetGetVarFunc(filters []expression.Expression) ([]expression.Expressio
...
 	canBePushDown, canNotBePushDown := splitSetGetVarFunc(p.Conditions)
-	retConditions, child := p.children[0].PredicatePushDown(append(canBePushDown, predicates...))
-	retConditions = append(retConditions, canNotBePushDown...)
-	if len(retConditions) > 0 {
-		p.Conditions = expression.PropagateConstant(p.ctx, retConditions)
-		// Return table dual when filter is constant false or null.
-		dual := Conds2TableDual(p, p.Conditions)
+	canBePushDown = append(canBePushDown, predicates...)
+	if len(canNotBePushDown) == 0 {
+		dual := Conds2TableDual(p, expression.PropagateConstant(p.ctx, canBePushDown))
```
[the remark this answers is not in the document]
Comment (gauss1314) [reply planner/core/rule_predicate_push_down.go]: Hi, now the return is 1 .
`set @p0 = '1', @p1 = '1';`
`execute stmt using @p0, @p1; `
Is that right?

**Comment by gauss1314** [comment]
Hi @francis0407 , according to the CI, there is a panic in another test `TestInspectionRuleTableExtractor`:
`PANIC: memtable_predicate_extractor_test.go:1005: extractorSuite.TestInspectionRuleTableExtractor
... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0xEEF9B1)`

I have no idea what is `InspectionRuleTableExtractor`'s relationship with `constant propagation for PredicatePushDown`

**Comment by francis0407** [comment]
Hi @gauss1314 ,
I think it may be similar to the tests that you have fixed in this PR. Some tests cases assume that the conditions like `a = 1 and a = 2` can be pushed down into the `LogicalMemtable`.

File: planner/core/rule_predicate_push_down.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (francis0407) [reply planner/core/rule_predicate_push_down.go]: @eurekaka How about call `PropagateConstant` in `DataSource.PredicatePushDown` and `Selection.PredicatePushDown`(only for conditions that cannot be pushed down).

**Comment by zz-jason** [comment]
I'm going to close this PR due to the issue has been fixed.
